pkgs/desktops/gnome/extensions/update-extensions.py

```diff
@@ -0,0 +1,281 @@
+            package_name_registry[shell_version][pname] += [uuid]
+        else:
+            package_name_registry[shell_version][pname] = [uuid]
+
+    return {
+        "uuid": uuid,
+        "name": sanitize_string(extension["name"]),
+        "pname": pname,
+        "description": sanitize_string(extension["description"]),
+        "link": "https://extensions.gnome.org" + extension["link"],
+        "shell_version_map": shell_version_map,
+    }
+
+
+def fetch_extensions() -> List[Dict[str, Any]]:
+    """
+    Fetch the extensions. The extensions database is scraped by using the search query. We simply go over it
+    page by page until we hit a non-full page or a 404 error.
+
+    The returned list is sorted by the age of the extension, in order to be deterministic.
+    """
+    page = 0
+    extensions = []
+    while True:
+        page += 1
+        logging.info("Scraping page " + str(page))
+        try:
+            with urllib.request.urlopen(
+                f"https://extensions.gnome.org/extension-query/?n_per_page=25&page={page}"
+            ) as url:
+                data = json.loads(url.read().decode())["extensions"]
+                responseLength = len(data)
+
+                for extension in data:
+                    extensions += [extension]
+
+                # If our page isn't "full", it must have been the last one
+                if responseLength < 25:
+                    logging.debug(
+                        f"\tThis page only has {responseLength} entries, so it must be the last one."
+                    )
+                    break
+        except urllib.error.HTTPError as e:
+            if e.code == 404:
+                # We reached past the last page and are done now
+                break
+            else:
+                raise
+
+    extensions.sort(key=itemgetter("pk"))
+    return extensions
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.DEBUG)
+
+    raw_extensions = fetch_extensions()
+
+    logging.info(f"Downloaded {len(raw_extensions)} extensions. Processing …")
+    processed_extensions: List[Dict[str, Any]] = []
+    for num, raw_extension in enumerate(raw_extensions):
+        processed_extension = process_extension(raw_extension)
+        if processed_extension:
+            processed_extensions += [processed_extension]
+            logging.debug(f"Processed {num + 1} / {len(raw_extensions)}")
+
+    logging.info(
+        f"Done. Writing results to extensions.json ({len(processed_extensions)} extensions in total)"
+    )
+
+    with open("extensions.json", "w") as out:
+        # Manually pretty-print the outer level, but then do one compact line per extension
+        # This allows for the diffs to be manageable (one line of change per extension) despite their quantity
+        for index, extension in enumerate(processed_extensions):
+            if index == 0:
+                out.write("[   ")
```

Comment:
Why this spacing after `[` and `,`?

Reply:
To indent the following content of that line (`json.dump(extension, out)`). See the `extensions.json` for how it looks, that should hopefully make it clear.

Reply:
Saw that but it still is not clear to me why – if you indent all lines, it is as if you did not indent any.

Reply:
The idea is that the brackets `[]` are on the topmost level, without indentation. The items then are indented one level; four spaces. As per the usual style for lists without trailing commas, the commas are on the first level, and then comes the content with the indentation.

Reply:
Yeah, that is what is weird. I would not consider the items with leading commas indented, there is only one space after the comma, which is consistent with typographic rules.

Reply:
So shall I change this to a single space (`[ `, `, `)? Or how would you like to see the formatting?

Reply:
Yeah, that would probably be for the best. (Sorry for this nit picking.)

Reply:
Done.